### PR TITLE
feat(number-field): add new event up-click and down-click

### PR DESCRIPTION
**Pull request by Theeraphat-Sorasetsakul**
## Description

- Add new event named up-click and down-click.
- Calls when tap at spinner.
- if do preventDefault, value-changed event should not call.

Fixes # ([ELF-2222](https://jira.refinitiv.com/browse/ELF-2222))

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: packages/elements/src/number-field/index.ts [hunk 1]
```diff
@@ -31,7 +31,9 @@ enum Direction {
  *
  * @fires value-changed - Fired when user commits a value change. The event is not triggered if `value` property is changed programmatically.
  * @fires error-changed - Fired when user inputs invalid value. The event is not triggered if `error` property is changed programmatically.
- *
+ * @fires up-clicked - Fired when user taps the spinner up. The event is not triggered if the spinner up is called programmatically.
```
Comment (Nantawat-Poothong) [inline packages/elements/src/number-field/index.ts]:
```suggestion
 * @fires up-clicked - Fired when user tap the spinner up. The event is not triggered if the spinner up is called programmatically.
```

File: packages/elements/src/number-field/index.ts [hunk 2]
```diff
@@ -31,7 +31,9 @@ enum Direction {
...
+ * @fires down-clicked - Fired when user taps the spinner down. The event is not triggered if the spinner down is called programmatically.
```
Comment (Nantawat-Poothong) [inline packages/elements/src/number-field/index.ts]:
```suggestion
 * @fires down-clicked - Fired when user tap the spinner down. The event is not triggered if the spinner down is called programmatically.
```

File: packages/elements/src/number-field/index.ts [hunk 3]
```diff
@@ -330,12 +332,29 @@ export class NumberField extends FormFieldElement {
...
+      );
+    } else if (direction === Direction.Down) {
+      event = this.dispatchEvent(
+        new CustomEvent('down-clicked', {
+          cancelable: true
+        })
+      );
+    }
+
+    if (event) {
```
Comment (Nantawat-Poothong) [inline packages/elements/src/number-field/index.ts:349]: There are such a use case that event will be false?

File: packages/elements/src/number-field/__demo__/index.html [hunk 4]
```diff
@@ -165,6 +165,46 @@
       </script>
     </demo-block>
 
+    <demo-block layout="normal" header="Up clicked event and Down clicked event" tags="event">
+      <ef-number-field id="up-down-clicked" placeholder="up-clicked and down-clicked"></ef-number-field>
+      <div>Event calls: <code id="up-down-clicked-event-text"></code></div>
```
Comment (Nantawat-Poothong) [inline packages/elements/src/number-field/__demo__/index.html]:
```suggestion
      <div>Event called: <code id="up-down-clicked-event-text"></code></div>
```

File: packages/elements/src/number-field/index.ts (hunk 3 above)
Comment (Nantawat-Poothong) [reply packages/elements/src/number-field/index.ts:349]: I got it now, it related to `preventDefault()`

File: packages/elements/src/number-field/index.ts [hunk 5]
```diff
@@ -330,12 +332,29 @@ export class NumberField extends FormFieldElement {
...
    */
   protected onApplyStep(direction: Direction): void {
-    try {
-      this.applyStepDirection(direction);
-      this.setSilentlyValueAndNotify();
-    } catch (error) {
-      // According to specs stepDown/stepUp may fail for some invalid inputs
-      // do nothing and report nothing in that case
+    let event = false;
+    if (direction === Direction.Up) {
```
Comment (Nantawat-Poothong) [inline packages/elements/src/number-field/index.ts]: How about this one?
```typescript
const eventName = direction === Direction.Up ? "up-clicked" : "down-clicked";
const event = this.dispatchEvent(
  new CustomEvent(eventName, {
    cancelable: true,
  })
);
```

File: packages/elements/src/number-field/index.ts (hunk 2 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/number-field/index.ts]: I think my grammar is correct. Can you explain about your suggestion?

File: packages/elements/src/number-field/index.ts (hunk 5 above)
Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/number-field/index.ts]: Updated from your suggestion.

File: packages/elements/src/number-field/index.ts (hunk 2 above)
Comment (Nantawat-Poothong) [reply packages/elements/src/number-field/index.ts]: I have no idea @wattachai-lseg  please help :D

Comment (wattachai-lseg) [reply packages/elements/src/number-field/index.ts]: I would go for `taps`.

**Comment by wattachai-lseg** [comment]
@Theeraphat-Sorasetsakul, are you sure about `down-clicked` and `up-clicked`? What about `step-up` & `step-down`?

**Comment by Theeraphat-Sorasetsakul** [comment]
Do you mean keydown ArrowUp and ArrowDown? If it is, yes `down-clicked` and `up-clicked` should be fired. @wattachai-lseg

**Comment by Nantawat-Poothong** [comment]
> Do you mean keydown ArrowUp and ArrowDown? If it is, yes `down-clicked` and `up-clicked` should be fired. @wattachai-lseg

I think he mean event naming.

**Comment by Theeraphat-Sorasetsakul** [comment]
> > Do you mean keydown ArrowUp and ArrowDown? If it is, yes `down-clicked` and `up-clicked` should be fired. @wattachai-lseg
> 
> I think he mean event naming.

But it said `up-clicked` and `down-clicked` in JIRA. Where do I find the reference of `step-up` and `step-down`? Sound like I miss something :fearful:.

**Comment by Theeraphat-Sorasetsakul** [comment]
After discussed, we agreed to change from `up-clicked` to `up-click` and `down-clicked` to `down-click`

**Comment by Theeraphat-Sorasetsakul** [comment]
After asked P'Bank, He said the events should not be fired by pressing arrow. So I will make a change.

File: packages/elements/src/number-field/index.ts [hunk 6]
```diff
@@ -316,12 +318,24 @@ export class NumberField extends FormFieldElement {
...
+    let direction = Direction.Down;
+    let eventName = '';
 
     if (target === this.spinnerDownEl) {
-      this.onApplyStep(Direction.Down);
+      eventName = 'down-click';
     } else if (target === this.spinnerUpEl) {
-      this.onApplyStep(Direction.Up);
+      eventName = 'up-click';
+      direction = Direction.Up;
```
Comment (Nantawat-Poothong) [inline packages/elements/src/number-field/index.ts]: I think we should revise this method to reading easier.

Comment (Nantawat-Poothong) [reply packages/elements/src/number-field/index.ts]: How about this one? I'm not sure is it more complex or not.

```typescript
private dispatchSpinnerClickEvent(eventName: string): boolean {
  return this.dispatchEvent(
    new CustomEvent(eventName, {
      cancelable: true
    })
  );
}

protected onSpinnerTap(event: TapEvent): void {
  if (this.disabled || this.readonly || event.defaultPrevented) {
    return;
  }

  const target = event.target;
  if (target === this.spinnerDownEl) {
    this.dispatchSpinnerClickEvent('down-click') && this.onApplyStep(Direction.Down);
  } else if (target === this.spinnerUpEl) {
    this.dispatchSpinnerClickEvent('up-click') && this.onApplyStep(Direction.Up);
  }
}
```

Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/number-field/index.ts]: That great and easier to read. Updated by your suggestion.

**Review by Nantawat-Poothong — DISMISSED** [review]
LGTM

File: packages/elements/src/number-field/index.ts [hunk 7]
```diff
@@ -31,6 +31,8 @@ enum Direction {
  *
  * @fires value-changed - Fired when user commits a value change. The event is not triggered if `value` property is changed programmatically.
  * @fires error-changed - Fired when user inputs invalid value. The event is not triggered if `error` property is changed programmatically.
+ * @fires step-up - Fired when user acts value up on both pressing arrow up or tapping the spinner up. The event is not triggered if the spinner up is called programmatically.
+ * @fires step-down - Fired when user acts value down on both pressing arrow down or tapping the spinner up. The event is not triggered if the spinner down is called programmatically.
```
Comment (Theeraphat-Sorasetsakul) [inline packages/elements/src/number-field/index.ts]:
```suggestion
 * @fires step-down - Fired when user acts value down on both pressing arrow down or tapping the spinner down. The event is not triggered if the spinner down is called programmatically.
```

File: packages/elements/src/number-field/index.ts [hunk 8]
```diff
@@ -305,6 +307,20 @@ export class NumberField extends FormFieldElement {
...
   }
 
+  /**
+   * Trigger step-up or step-down event and return the event is cancelled
+   * @param direction Up or Down
+   * @returns {boolean} cancelled event
+   */
+  private dispatchStepEvent(direction: Direction): boolean {
+    const eventName = direction === Direction.Up ? 'step-up' : 'step-down';
+    return this.dispatchEvent(
```
Comment (Nantawat-Poothong) [inline packages/elements/src/number-field/index.ts:317]: is it correct if this method return `true` = cancelled event?

Comment (Theeraphat-Sorasetsakul) [reply packages/elements/src/number-field/index.ts:317]: You're right. I see wrong description.  Updated a new description.

**Review by Nantawat-Poothong — APPROVED** [review]
LGTM

**Comment by bualoy-napat** [comment]
LGTM

**Pull request merged**